// react-error-overlay/lib/containers/RuntimeErrorContainer.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

import React, { PureComponent } from 'react';
import ErrorOverlay from '../components/ErrorOverlay';
import CloseButton from '../components/CloseButton';
import NavigationBar from '../components/NavigationBar';
import RuntimeError from './RuntimeError';
import Footer from '../components/Footer';

var RuntimeErrorContainer = function (_PureComponent) {
  _inherits(RuntimeErrorContainer, _PureComponent);

  function RuntimeErrorContainer() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, RuntimeErrorContainer);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = RuntimeErrorContainer.__proto__ || Object.getPrototypeOf(RuntimeErrorContainer)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      currentIndex: 0
    }, _this.previous = function () {
      _this.setState(function (state, props) {
        return {
          currentIndex: state.currentIndex > 0 ? state.currentIndex - 1 : props.errorRecords.length - 1
        };
      });
    }, _this.next = function () {
      _this.setState(function (state, props) {
        return {
          currentIndex: state.currentIndex < props.errorRecords.length - 1 ? state.currentIndex + 1 : 0
        };
      });
    }, _this.shortcutHandler = function (key) {
      if (key === 'Escape') {
        _this.props.close();
      } else if (key === 'ArrowLeft') {
        _this.previous();
      } else if (key === 'ArrowRight') {
        _this.next();
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(RuntimeErrorContainer, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          errorRecords = _props.errorRecords,
          close = _props.close;

      var totalErrors = errorRecords.length;
      return React.createElement(
        ErrorOverlay,
        { shortcutHandler: this.shortcutHandler },
        React.createElement(CloseButton, { close: close }),
        totalErrors > 1 && React.createElement(NavigationBar, {
          currentError: this.state.currentIndex + 1,
          totalErrors: totalErrors,
          previous: this.previous,
          next: this.next
        }),
        React.createElement(RuntimeError, {
          errorRecord: errorRecords[this.state.currentIndex],
          launchEditorEndpoint: this.props.launchEditorEndpoint
        }),
        React.createElement(Footer, {
          line1: 'This screen is visible only in development. It will not appear if the app crashes in production.',
          line2: 'Open your browser\u2019s developer console to further inspect this error.'
        })
      );
    }
  }]);

  return RuntimeErrorContainer;
}(PureComponent);

export default RuntimeErrorContainer;
