// react-error-overlay/lib/containers/StackFrame.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

import React, { Component } from 'react';
import CodeBlock from './StackFrameCodeBlock';
import { getPrettyURL } from '../utils/getPrettyURL';
import { darkGray } from '../styles';

var linkStyle = {
  fontSize: '0.9em',
  marginBottom: '0.9em'
};

var anchorStyle = {
  textDecoration: 'none',
  color: darkGray,
  cursor: 'pointer'
};

var codeAnchorStyle = {
  cursor: 'pointer'
};

var toggleStyle = {
  marginBottom: '1.5em',
  color: darkGray,
  cursor: 'pointer',
  border: 'none',
  display: 'block',
  width: '100%',
  textAlign: 'left',
  background: '#fff',
  fontFamily: 'Consolas, Menlo, monospace',
  fontSize: '1em',
  padding: '0px',
  lineHeight: '1.5'
};

var StackFrame = function (_Component) {
  _inherits(StackFrame, _Component);

  function StackFrame() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, StackFrame);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = StackFrame.__proto__ || Object.getPrototypeOf(StackFrame)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      compiled: false
    }, _this.toggleCompiled = function () {
      _this.setState(function (state) {
        return {
          compiled: !state.compiled
        };
      });
    }, _this.openInEditor = function () {
      if (!_this.canOpenInEditor()) {
        return;
      }
      var _this$props$frame = _this.props.frame,
          sourceFileName = _this$props$frame._originalFileName,
          sourceLineNumber = _this$props$frame._originalLineNumber;
      // Keep this in sync with react-error-overlay/middleware.js

      fetch(_this.props.launchEditorEndpoint + '?fileName=' + window.encodeURIComponent(sourceFileName) + '&lineNumber=' + window.encodeURIComponent(sourceLineNumber || 1)).then(function () {}, function () {});
    }, _this.onKeyDown = function (e) {
      if (e.key === 'Enter') {
        _this.openInEditor();
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(StackFrame, [{
    key: 'canOpenInEditor',
    value: function canOpenInEditor() {
      if (!this.props.launchEditorEndpoint) {
        return;
      }
      var sourceFileName = this.props.frame._originalFileName;
      // Unknown file

      if (!sourceFileName) {
        return false;
      }
      // e.g. "/path-to-my-app/webpack/bootstrap eaddeb46b67d75e4dfc1"
      var isInternalWebpackBootstrapCode = sourceFileName.trim().indexOf(' ') !== -1;
      if (isInternalWebpackBootstrapCode) {
        return false;
      }
      // Code is in a real file
      return true;
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          frame = _props.frame,
          contextSize = _props.contextSize,
          critical = _props.critical,
          showCode = _props.showCode;
      var fileName = frame.fileName,
          lineNumber = frame.lineNumber,
          columnNumber = frame.columnNumber,
          scriptLines = frame._scriptCode,
          sourceFileName = frame._originalFileName,
          sourceLineNumber = frame._originalLineNumber,
          sourceColumnNumber = frame._originalColumnNumber,
          sourceLines = frame._originalScriptCode;

      var functionName = frame.getFunctionName();

      var compiled = this.state.compiled;
      var url = getPrettyURL(sourceFileName, sourceLineNumber, sourceColumnNumber, fileName, lineNumber, columnNumber, compiled);

      var codeBlockProps = null;
      if (showCode) {
        if (compiled && scriptLines && scriptLines.length !== 0 && lineNumber != null) {
          codeBlockProps = {
            lines: scriptLines,
            lineNum: lineNumber,
            columnNum: columnNumber,
            contextSize: contextSize,
            main: critical
          };
        } else if (!compiled && sourceLines && sourceLines.length !== 0 && sourceLineNumber != null) {
          codeBlockProps = {
            lines: sourceLines,
            lineNum: sourceLineNumber,
            columnNum: sourceColumnNumber,
            contextSize: contextSize,
            main: critical
          };
        }
      }

      var canOpenInEditor = this.canOpenInEditor();
      return React.createElement(
        'div',
        null,
        React.createElement(
          'div',
          null,
          functionName
        ),
        React.createElement(
          'div',
          { style: linkStyle },
          React.createElement(
            'a',
            {
              style: canOpenInEditor ? anchorStyle : null,
              onClick: canOpenInEditor ? this.openInEditor : null,
              onKeyDown: canOpenInEditor ? this.onKeyDown : null,
              tabIndex: canOpenInEditor ? '0' : null
            },
            url
          )
        ),
        codeBlockProps && React.createElement(
          'span',
          null,
          React.createElement(
            'a',
            {
              onClick: canOpenInEditor ? this.openInEditor : null,
              style: canOpenInEditor ? codeAnchorStyle : null
            },
            React.createElement(CodeBlock, codeBlockProps)
          ),
          React.createElement(
            'button',
            { style: toggleStyle, onClick: this.toggleCompiled },
            'View ' + (compiled ? 'source' : 'compiled')
          )
        )
      );
    }
  }]);

  return StackFrame;
}(Component);

export default StackFrame;
