// react-error-overlay/lib/listenToRuntimeErrors.js
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

import { register as registerError, unregister as unregisterError } from './effects/unhandledError';
import { register as registerPromise, unregister as unregisterPromise } from './effects/unhandledRejection';
import { register as registerStackTraceLimit, unregister as unregisterStackTraceLimit } from './effects/stackTraceLimit';
import { permanentRegister as permanentRegisterConsole, registerReactStack, unregisterReactStack } from './effects/proxyConsole';
import { massage as massageWarning } from './utils/warnings';
import getStackFrames from './utils/getStackFrames';

var CONTEXT_SIZE = 3;

export function listenToRuntimeErrors(crash) {
  var filename = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '/static/js/bundle.js';

  function crashWithFrames(error) {
    var unhandledRejection = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

    getStackFrames(error, unhandledRejection, CONTEXT_SIZE).then(function (stackFrames) {
      if (stackFrames == null) {
        return;
      }
      crash({
        error: error,
        unhandledRejection: unhandledRejection,
        contextSize: CONTEXT_SIZE,
        stackFrames: stackFrames
      });
    }).catch(function (e) {
      console.log('Could not get the stack frames of error:', e);
    });
  }
  registerError(window, function (error) {
    return crashWithFrames(error, false);
  });
  registerPromise(window, function (error) {
    return crashWithFrames(error, true);
  });
  registerStackTraceLimit();
  registerReactStack();
  permanentRegisterConsole('error', function (warning, stack) {
    var data = massageWarning(warning, stack);
    crashWithFrames(
    // $FlowFixMe
    {
      message: data.message,
      stack: data.stack,
      __unmap_source: filename
    }, false);
  });

  return function stopListening() {
    unregisterStackTraceLimit();
    unregisterPromise(window);
    unregisterError(window);
    unregisterReactStack();
  };
}
