// react-error-overlay/lib/components/CodeBlock.js
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

import React from 'react';
import { redTransparent, yellowTransparent } from '../styles';

var _preStyle = {
  display: 'block',
  padding: '0.5em',
  marginTop: '0.5em',
  marginBottom: '0.5em',
  overflowX: 'auto',
  whiteSpace: 'pre-wrap',
  borderRadius: '0.25rem'
};

var primaryPreStyle = Object.assign({}, _preStyle, {
  backgroundColor: redTransparent
});

var secondaryPreStyle = Object.assign({}, _preStyle, {
  backgroundColor: yellowTransparent
});

var codeStyle = {
  fontFamily: 'Consolas, Menlo, monospace'
};

function CodeBlock(props) {
  var preStyle = props.main ? primaryPreStyle : secondaryPreStyle;
  var codeBlock = { __html: props.codeHTML };

  return React.createElement(
    'pre',
    { style: preStyle },
    React.createElement('code', { style: codeStyle, dangerouslySetInnerHTML: codeBlock })
  );
}

export default CodeBlock;
