// react-error-overlay/lib/components/Collapsible.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

import React, { Component } from 'react';
import { black } from '../styles';

var _collapsibleStyle = {
  color: black,
  cursor: 'pointer',
  border: 'none',
  display: 'block',
  width: '100%',
  textAlign: 'left',
  background: '#fff',
  fontFamily: 'Consolas, Menlo, monospace',
  fontSize: '1em',
  padding: '0px',
  lineHeight: '1.5'
};

var collapsibleCollapsedStyle = Object.assign({}, _collapsibleStyle, {
  marginBottom: '1.5em'
});

var collapsibleExpandedStyle = Object.assign({}, _collapsibleStyle, {
  marginBottom: '0.6em'
});

var Collapsible = function (_Component) {
  _inherits(Collapsible, _Component);

  function Collapsible() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Collapsible);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Collapsible.__proto__ || Object.getPrototypeOf(Collapsible)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      collapsed: true
    }, _this.toggleCollaped = function () {
      _this.setState(function (state) {
        return {
          collapsed: !state.collapsed
        };
      });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Collapsible, [{
    key: 'render',
    value: function render() {
      var count = this.props.children.length;
      var collapsed = this.state.collapsed;
      return React.createElement(
        'div',
        null,
        React.createElement(
          'button',
          {
            onClick: this.toggleCollaped,
            style: collapsed ? collapsibleCollapsedStyle : collapsibleExpandedStyle
          },
          (collapsed ? '▶' : '▼') + (' ' + count + ' stack frames were ') + (collapsed ? 'collapsed.' : 'expanded.')
        ),
        React.createElement(
          'div',
          { style: { display: collapsed ? 'none' : 'block' } },
          this.props.children,
          React.createElement(
            'button',
            {
              onClick: this.toggleCollaped,
              style: collapsibleExpandedStyle
            },
            '\u25B2 ' + count + ' stack frames were expanded.'
          )
        )
      );
    }
  }]);

  return Collapsible;
}(Component);

export default Collapsible;
