// react-error-overlay/lib/components/ErrorOverlay.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

import React, { Component } from 'react';
import { black } from '../styles';

var overlayStyle = {
  position: 'relative',
  display: 'inline-flex',
  flexDirection: 'column',
  height: '100%',
  width: '1024px',
  maxWidth: '100%',
  overflowX: 'hidden',
  overflowY: 'auto',
  padding: '0.5rem',
  boxSizing: 'border-box',
  textAlign: 'left',
  fontFamily: 'Consolas, Menlo, monospace',
  fontSize: '11px',
  whiteSpace: 'pre-wrap',
  wordBreak: 'break-word',
  lineHeight: 1.5,
  color: black
};

var ErrorOverlay = function (_Component) {
  _inherits(ErrorOverlay, _Component);

  function ErrorOverlay() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ErrorOverlay);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ErrorOverlay.__proto__ || Object.getPrototypeOf(ErrorOverlay)).call.apply(_ref, [this].concat(args))), _this), _this.iframeWindow = null, _this.getIframeWindow = function (element) {
      if (element) {
        var document = element.ownerDocument;
        _this.iframeWindow = document.defaultView;
      }
    }, _this.onKeyDown = function (e) {
      var shortcutHandler = _this.props.shortcutHandler;

      if (shortcutHandler) {
        shortcutHandler(e.key);
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ErrorOverlay, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      window.addEventListener('keydown', this.onKeyDown);
      if (this.iframeWindow) {
        this.iframeWindow.addEventListener('keydown', this.onKeyDown);
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      window.removeEventListener('keydown', this.onKeyDown);
      if (this.iframeWindow) {
        this.iframeWindow.removeEventListener('keydown', this.onKeyDown);
      }
    }
  }, {
    key: 'render',
    value: function render() {
      return React.createElement(
        'div',
        { style: overlayStyle, ref: this.getIframeWindow },
        this.props.children
      );
    }
  }]);

  return ErrorOverlay;
}(Component);

export default ErrorOverlay;
